pkg/kubelet/cm/memorymanager/policy_single_numa.go

```diff
@@ -357,44 +361,76 @@ func (p *singleNUMAPolicy) calculateHints(s state.State, requestedResources map[
 					totalFreeSize[resourceName] = 0
 				}
 				totalFreeSize[resourceName] += machineState[nodeId].MemoryMap[resourceName].Free
+
+				if _, ok := totalAllocatableSize[resourceName]; !ok {
+					totalAllocatableSize[resourceName] = 0
+				}
+				totalAllocatableSize[resourceName] += machineState[nodeId].MemoryMap[resourceName].Allocatable
+			}
+		}
+
+		// verify that for all memory types the node mask has enough allocatable resources
+		for resourceName, requestedSize := range requestedResources {
+			if totalAllocatableSize[resourceName] < requestedSize {
+				return
 			}
 		}
 
-		// verify that for all memory types the node mask has enough resources
+		// set the minimum amount of NUMA nodes that can satisfy the container resources requests
+		if mask.Count() < minAffinitySize {
+			minAffinitySize = mask.Count()
+		}
+
+		// verify that for all memory types the node mask has enough free resources
 		for resourceName, requestedSize := range requestedResources {
 			if totalFreeSize[resourceName] < requestedSize {
 				return
 			}
 		}
 
-		preferred := p.isHintPreferred(maskBits)
 		// add the node mask as topology hint for all memory types
 		for resourceName := range requestedResources {
 			if _, ok := hints[string(resourceName)]; !ok {
 				hints[string(resourceName)] = []topologymanager.TopologyHint{}
 			}
 			hints[string(resourceName)] = append(hints[string(resourceName)], topologymanager.TopologyHint{
 				NUMANodeAffinity: mask,
-				Preferred:        preferred,
+				Preferred:        false,
 			})
 		}
 	})
 
+	// update hints preferred according to multiNUMAGroups, in case when it wasn't provided, the default
+	// behaviour to prefer the minimal amount of NUMA nodes will be used
+	for resourceName := range requestedResources {
+		for i, hint := range hints[string(resourceName)] {
+			hints[string(resourceName)][i].Preferred = p.isHintPreferred(hint.NUMANodeAffinity.GetBits(), minAffinitySize)
+		}
+	}
+
 	return hints
 }
 
-func (p *singleNUMAPolicy) isHintPreferred(maskBits []int) bool {
+func (p *singleNUMAPolicy) isHintPreferred(maskBits []int, minAffinitySize int) bool {
+	if len(p.multiNUMAGroups) == 0 {
+		if len(maskBits) == minAffinitySize {
+			return true
+		}
+
+		return false
+	}
 	// check if the mask is for the single NUMA node hint
 	if len(maskBits) == 1 {
-		// the node should be used only for multiple NUMA memory allocation
-		return !p.isMultipleNUMANode(maskBits[0])
+		// the node should be used only for multi NUMA memory allocation
+		return !p.isMultiNUMANode(maskBits[0])
 	}
 
-	return p.isMultipleNUMAGroup(maskBits)
+	return p.isMultiNUMAGroup(maskBits)
```

Comment:
I think we should also check here if a group is a minimal possible hint to set it as preferred.
```
if len(maskBits) == minAffinitySize {...}
```

I understand it as following logic:
we have:

Case 1):
Memory allocatable: 0: [[10GB, 10GB, 10Gb, 10GB]
NUMA groups: [2,3]
Requirement: 5GB
Possible hints [0001-preferred,0010-preferred,1100]

Case 2):
Memory allocatable: 0: [[10GB, 10GB, 10Gb, 10GB]
NUMA groups: [2,3]
Requirement: 15GB
Possible hints [1100-preferred]